src/v/datalake/protobuf_to_arrow_converter.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/protobuf_to_arrow_converter.h"
+
+#include "datalake/errors.h"
+#include "datalake/proto_to_arrow_struct.h"
+
+#include <arrow/type.h>
+
+#include <memory>
+#include <stdexcept>
+
+namespace datalake {
+
+proto_to_arrow_converter::proto_to_arrow_converter(const ss::sstring& schema) {
+    initialize_protobuf_schema(schema);
+    if (!initialize_struct_converter()) {
+        throw initialization_error("Could not initialize arrow arrays");
+    }
+}
+arrow_converter_status
+proto_to_arrow_converter::add_message(const ss::sstring& serialized_message) {
+    std::unique_ptr<google::protobuf::Message> message = parse_message(
+      serialized_message);
+    if (message == nullptr) {
+        return arrow_converter_status::parse_error;
+    }
+
+    if (!_struct_converter->add_top_level_message(message.get()).ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+arrow_converter_status proto_to_arrow_converter::finish_batch() {
+    if (!_struct_converter->finish_batch().ok()) {
+        return arrow_converter_status::internal_error;
+    }
+    return arrow_converter_status::ok;
+}
+std::shared_ptr<arrow::Table> proto_to_arrow_converter::build_table() {
+    auto table_result = arrow::Table::FromChunkedStructArray(
+      _struct_converter->finish());
```

Comment:
> Arrow does attempt to avoid large contiguous allocations

I think we need to understand the underlying allocation strategy. For example if I have a  ChunkedArray with 100KiB with of data and add one more element, does it double the capacity? We can easily get oversized allocations if data structures are on the hot path and aren't using one of our optimized structures (chunked_vector, chunked_hash_map, chunked_fifo, etc). Nothing to do now, but I expect us to either have some really amazing in depth automated tests here or rewrite it ourselves to use our own datastructures and be seastar scheduler friendly (which will also require amazing in-depth automated tests 😄) before supporting arbitrary customer workloads.

Reply:
Another note is that the chunking method for this in memory data structure is the same chunking as the written parquet file, and I am sure that we will want the ability write larger chunks in the parquet file for downstream system performance.